src/clib/pio_msg.c

```diff
@@ -17,20 +18,1061 @@ extern int my_rank;
 extern int pio_log_level;
 #endif /* PIO_ENABLE_LOGGING */
 
-/** This function is run on the IO tasks to handle nc_inq_type*()
+/* A global MPI communicator for waiting on async I/O service messages */
+MPI_Comm pio_async_service_msg_comm = MPI_COMM_NULL;
+
+/* Create the global MPI communicator for async service messaging.
+ * This communicator is used by the async I/O service message handler to
+ * communicate the iosystem id between the I/O procs. It is a dup of the
+ * io_comm
+ * @param io_comm The I/O communicator used for creating the async service
+ * message communicator
+ * @param msg_comm Pointer to the newly created (global) asynchronous I/O
+ * message communicator
+ * @returns PIO_NOERR on success, a pio error code otherwise
+ */
+int create_async_service_msg_comm(const MPI_Comm io_comm, MPI_Comm *msg_comm)
+{
+    int ret = PIO_NOERR;
+
+    assert(msg_comm && (pio_async_service_msg_comm == MPI_COMM_NULL));
+
+    LOG((2, "Creating global async I/O service msg comm"));
+    *msg_comm = MPI_COMM_NULL;
+
+    if(io_comm != MPI_COMM_NULL)
+    {
+        ret = MPI_Comm_dup(io_comm, &pio_async_service_msg_comm);
+        if(ret != MPI_SUCCESS)
+        {
+            return check_mpi(NULL, NULL, ret, __FILE__, __LINE__);
+        }
+        *msg_comm = pio_async_service_msg_comm;
+    }
+    return ret;
+}
+
+/* Delete/free the global MPI communicator used for asynchronous I/O service
+ * messaging
+ */
+void delete_async_service_msg_comm(void )
+{
+    if(pio_async_service_msg_comm != MPI_COMM_NULL)
+    {
+        LOG((2, "Deleting global async I/O service msg comm"));
+        MPI_Comm_free(&pio_async_service_msg_comm);
+        pio_async_service_msg_comm = MPI_COMM_NULL;
+    }
+}
+
+char pio_async_msg_sign[PIO_MAX_MSGS][PIO_MAX_ASYNC_MSG_ARGS];
+int init_async_msgs_sign(void )
```

Comment:
I am not attracted to this approach at all. In fact, it seems that it makes the code more complex, more places to get things wrong, harder to debug, and still adds ~1K LOC to the file. So how is this an improvement?

The handling of async messages is similar but not the same for all functions in the API. There will never be more API functions. So creating code to handle messages in this general way is not helpful. It might be if there were thousands of messages, or if new messages were being added all the time. But for this situation, it seems like a lot more code than necessary to meet requirements.

Extra code = extra cost.